modoboa/admin/lib.py

```diff
@@ -190,11 +190,52 @@ def get_dns_resolver():
     return resolver
 
 
+def get_authoritative_server(domain, resolver=None):
+    if not resolver:
+       resolver = get_dns_resolver()
+
+    dnsname = dns.name.from_text(domain)
+
+    while True:
+        try:
+            answers = dns.resolver.resolve(dnsname, 'NS')
+            if answers:
+                first_answer = answers[0]
+                if hasattr(first_answer, "target"):
+                    return str(first_answer.target)
+                if hasattr(first_answer, "address"):
+                    return str(first_answer.address)
+        except dns.resolver.NoAnswer as e:
+            dnsname = dnsname.parent()
+        else:
+            dnsname = dnsname.parent()
```

Comment:
@arthru What happens if parent is None?

Reply:
the ultimate parent is `.` that is not shown on most domains

A ns query on `.` answers to the root dns servers that always answers :

```
$ dig ns .

; <<>> DiG 9.18.27 <<>> ns .
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 58155
;; flags: qr rd ra ad; QUERY: 1, ANSWER: 13, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 1472
;; QUESTION SECTION:
;.				IN	NS

;; ANSWER SECTION:
.			74393	IN	NS	a.root-servers.net.
.			74393	IN	NS	b.root-servers.net.
.			74393	IN	NS	c.root-servers.net.
.			74393	IN	NS	d.root-servers.net.
.			74393	IN	NS	e.root-servers.net.
.			74393	IN	NS	f.root-servers.net.
.			74393	IN	NS	g.root-servers.net.
.			74393	IN	NS	h.root-servers.net.
.			74393	IN	NS	i.root-servers.net.
.			74393	IN	NS	j.root-servers.net.
.			74393	IN	NS	k.root-servers.net.
.			74393	IN	NS	l.root-servers.net.
.			74393	IN	NS	m.root-servers.net.

;; Query time: 10 msec
;; SERVER: 192.168.111.254#53(192.168.111.254) (UDP)
;; WHEN: Wed Jul 17 14:55:13 CEST 2024
;; MSG SIZE  rcvd: 239
```

so the parent should never be None

Reply:
Ok I see.